### ISSUE TITLE
Make collection definitions better

### ISSUE THREAD

**Issue by bradwilson** [xunit/xunit#2007]
Today, you can use `[Collection]` with a string without a collection definition, and when you need a collection definition, you have to know to create an empty class with `[CollectionDefinition]` on it. This is a terrible design from a usability standpoint.

New design: TBD.

**Comment by bradwilson** [comment]
If you have suggestions, please feel free to provide them here.

**Comment by amoerie** [comment]
I would vote to move towards interfaces instead of attributes. ITestCollection could be the interface, with settings and a name simply being the members of the interface. Interfaces are more convenient than attributes because it removes the constraint that all settings must be constants, and it removes the weird empty classes that only exist so you can put an attribute on them.

To declare that test classes belong to a collection, it can simply be a matter of implementing ITest<TCollection> where TCollection: ITestCollection.

I wouldn't provide a hard equivalent to the collection with a string attribute (without a collection definition), provided that implementing is ITestCollection is very quick and easy. (Either by keeping ITestCollection sparse, or by providing a Default test collection with sane defaults)

Assuming you considered this interface based design for the first design, why did the attribute based design win over this one? 

Disclaimer: I'm not familiar with the internals of xUnit, so I cannot vouch for the feasibility of the ideas I'm proposing.

**Comment by bradwilson** [comment]
>Assuming you considered this interface based design for the first design, why did the attribute based design win over this one?

This isn't a design we considered. It is an interesting one. The only thing that's sort of odd about it is that you can "legally" (from a language perspective) belong to two test collections, which is not legal in xUnit.net terms.

**Comment by amoerie** [comment]
Yes, it is technically possible to belong to two collections then, but the only way to prevent that is an abstract class to inherit from. However, inheriting from an abstract class just to provide some metadata feels wrong. It feels like abuse of the language.

Furthermore, there is precedent in xUnit throwing runtime exceptions for code that compiles but is not valid xUnit code, for example InlineData with the wrong number of arguments.

There's always xUnit.analyzers to assist people at design time, no?

**Comment by StevenLiekens** [comment]
> you have to know to create an empty class with [CollectionDefinition] on it.

Can you keep the CollectionAttribute but drop the CollectionDefinitionAttribute? Using CollectionAttribute sort of implies that a collection exists.

For collection fixtures, I wouldn't mind using attributes there as well.

``` csharp

[CollectionFixture("Database collection")]
public class DatabaseFixture : IDisposable
{
    // ...
}

[Collection("Database collection")]
public class DatabaseTestClass1
{
    DatabaseFixture fixture;

    public DatabaseTestClass1(DatabaseFixture fixture)
    {
        this.fixture = fixture;
    }
}

[Collection("Database collection")]
public class DatabaseTestClass2
{
    // ...
}
```

**Comment by jzabroski** [comment]
> Today, you can use `[Collection]` with a string without a collection definition, and when you need a collection definition, you have to know to create an empty class with `[CollectionDefinition]` on it. This is a terrible design from a usability standpoint.
> 
> New design: TBD.

Why not just implement https://github.com/xunit/xunit/issues/1277 ?

While I understand your point is about API Discoverability, creating an empty class isn't required, as far as I can tell.  This works fine:

```c#
using System.Threading;
using Xunit;

namespace XUnitTestProject
{
	[CollectionDefinition(nameof(UnitTest1), DisableParallelization = true)]
	[Collection(nameof(UnitTest1))]
	public class UnitTest1
	{
		[Fact]
		public void Test1()
		{
			NotThreadSafeResource.Use();
			Thread.Sleep(1000);
			NotThreadSafeResource.Release();
		}

		[Fact]
		public void Test3()
		{
			NotThreadSafeResource.Use();
			Thread.Sleep(1000);
			NotThreadSafeResource.Release();
		}
	}
}
```

**Comment by jzabroski** [comment]
Coming back to this 4 years later:

> the purpose of custom test collections (so you can have multiple classes that belong to it).

I think this is the core problem.  The documentation around Collections is confusing and unclear:

1. CollectionDefinition DisableParallelization defaults to false, but if you only have a Collection without a CollectionDefinition, it defaults to true.  
2. CollectionDefinition is primarily documented in the context of Collection Fixtures.

This is only meant as feedback as an end user on what I find non-intuitive. It's also not clear from the API (the Attributes make you think the default is that a collection _is_ parallelizable).

If I had to suggest an alternative, switch to using source generators to specify the "work" and "span" of a DAG of tests. Newer frameworks like TUnit are inspiring in this regard. Then problems like test discovery become simple partitioning of ICollection(s) of tests and test data providers (viz a viz IDataAttributes). Reifying test problem domain concepts in the context of work and span makes it easier to discuss how to optimize test execution and improve performance. Just a hunch, anyway.

I may think more about this going forward as I'm trying to squeeze performance out of a large suite of tests.  I was hoping I could bend xunit to my will more easily than I actually could.
